### PR TITLE
(PXP-5525): Add blank version of record

### PR DESCRIPTION
**Pull request by mpingram**
Jira Ticket: [PXP-5525](https://ctds-planx.atlassian.net/browse/PXP-5525)

Adds support for the POST `/index/blank/{{GUID}}` endpoint, which allows creation of a new blank version of an existing record.

### New Features
- Add support for creating a new, blank version of a record through the POST `/index/blank/{{GUID}}` endpoint.

### CODE REVIEW

File: indexd/index/blueprint.py [hunk 1]
```diff
@@ -295,6 +295,30 @@ def post_index_blank_record():
...
+    Only uploader and optionally file_name fields are filled.
+    Returns the GUID of the new blank version and the baseid common to all versions
+    of the record.
+    """
+    uploader = flask.request.get_json().get("uploader")
+    file_name = flask.request.get_json().get("file_name")
+    if not uploader:
+        raise UserError("no uploader specified")
+
+    # authorize done in add_blank_version for both the old and new authz
```
Comment (paulineribeyre) [inline indexd/index/blueprint.py:312]:
```suggestion
    # authorize done in add_blank_version for the existing record's authz
```

Comment (paulineribeyre) [inline indexd/index/blueprint.py:312]: i see this comment was copied from `add_index_record_version` and it actually looks like `add_version` is _not_ checking both old and new authz! that worries me

File: indexd/index/drivers/alchemy.py [hunk 2]
```diff
@@ -1149,6 +1149,42 @@ def add_version(
...
+            except NoResultFound:
+                raise NoRecordFound("no record found")
+            except MultipleResultsFound:
+                raise MultipleRecordsFound("multiple records found")
+
+            auth.authorize("update", [u.resource for u in old_record.authz])
+
+            new_record = IndexRecord()
+            did = str(uuid.uuid4())
+            if self.config.get("PREPEND_PREFIX"):
```
Comment (paulineribeyre) [inline indexd/index/drivers/alchemy.py:1171]: `add_version` for regular non-blank records is missing this `PREPEND_PREFIX` block :(

File: openapis/swagger.yaml [hunk 3]
```diff
@@ -245,6 +245,37 @@ paths:
...
+          required: true
+          type: string
+        - in: body
+          name: body
+          description: Metadata object that needs to be added to the store
+          required: true
+          schema:
+            $ref: '#/definitions/InputBlankInfo'
+      responses:
+        '200':
```
Comment (paulineribeyre) [inline openapis/swagger.yaml:271]: your endpoint returns 201 not 200

and looks like `POST /index/blank` is returning 201 too and the swagger says 200, if you can fix that as well that'd be appreciated 🙏

File: openapis/swagger.yaml [hunk 4]
```diff
@@ -1350,6 +1381,18 @@ definitions:
       rev:
         type: string
         pattern: '^[0-9a-f]{8}$'
+  OutputRefWithURL:
```
Comment (paulineribeyre) [inline openapis/swagger.yaml:1384]: what's this?

File: openapis/swagger.yaml (hunk 3 above)
Comment (mpingram) [reply openapis/swagger.yaml:271]: Good catch, thank you! I'll update `/index/blank` as well.

File: openapis/swagger.yaml (hunk 4 above)
Comment (mpingram) [reply openapis/swagger.yaml:1384]: Good catch, it's an orphan object schema that I made and ended up not needing. I'll delete it.

File: indexd/index/drivers/alchemy.py (hunk 2 above)
Comment (mpingram) [reply indexd/index/drivers/alchemy.py:1171]: Yeah 😬I could add it to add_version but I didn't want to break anything.

File: indexd/index/blueprint.py (hunk 1 above)
Comment (mpingram) [reply indexd/index/blueprint.py:312]: I see -- how big of a deal is that??

Comment (paulineribeyre) [reply indexd/index/blueprint.py:312]: sorry 🤦‍♀️ i checked the code again and it does check both. woops

Comment (mpingram) [reply indexd/index/blueprint.py:312]: haha no way! Ok phew ✨

File: indexd/index/drivers/alchemy.py (hunk 2 above)
Comment (mpingram) [reply indexd/index/drivers/alchemy.py:1171]: I think I'll add the prepend_prefix code to add_version and document it as a bugfix in this PR. What do you think?

File: tests/test_client.py [hunk 5]
```diff
@@ -576,6 +576,93 @@ def test_create_blank_record_with_file_name(client, user):
     assert_blank(rec)
 
 
+def test_create_blank_version(client, user):
```
[the remark this answers is not in the document]
Comment (mpingram) [reply tests/test_client.py:579]: That's a great question. I'll do a manual test and see

File: indexd/index/drivers/alchemy.py (hunk 2 above)
Comment (paulineribeyre) [reply indexd/index/drivers/alchemy.py:1171]: hey, i talked with other arcadians and i have a fix a for it, so don't worry about it
thanks :D

Comment (mpingram) [reply indexd/index/drivers/alchemy.py:1171]: Incredible, thank u!

File: indexd/index/drivers/alchemy.py [hunk 6]
```diff
@@ -1166,10 +1166,17 @@ def add_blank_version(self, current_did, file_name=None, uploader=None):
 
             auth.authorize("update", [u.resource for u in old_record.authz])
 
+            # handle the edgecase where new_did matches the original doc's did to
+            # prevent sqlalchemy FlushError
+            if new_did == old_record.did:
+                raise UserError("{did} already exists".format(did=new_did), 400)
```
Comment (paulineribeyre) [inline indexd/index/drivers/alchemy.py:1176]: does `except IntegrityError` later not catch this? :o

File: indexd/index/drivers/alchemy.py [hunk 7]
```diff
@@ -1172,7 +1218,7 @@ def get_all_versions(self, did):
                 else:
                     baseid = record.baseid
             except MultipleResultsFound:
-                raise MultipleRecordsFound("multiple records found")
+                raise MultipleRecordsFoundadd_version("multiple records found")
```
Comment (paulineribeyre) [inline indexd/index/drivers/alchemy.py:1221]: is this a typo?

File: tests/test_client.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (paulineribeyre) [reply tests/test_client.py:579]: thanks for the extra test!! so doing that works - is it what we want, design wise?

File: indexd/index/drivers/alchemy.py (hunk 7 above)
Comment (mpingram) [reply indexd/index/drivers/alchemy.py:1221]: 😮 Yes it is

File: tests/test_client.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mpingram) [reply tests/test_client.py:579]: To my mind this was the least surprising option, considering that the only things that get passed down when you make a new blank version of a record are the `did`, `baseid`, and `acl/authz` fields. What do u think?

**Pull request merged**